File: src/DTOs/Validation/ValidationRule.php
<?php

namespace ReactUbiquitous\DTOs\Validation;

use ReactUbiquitous\Contracts\SerializableInterface;

final class ValidationRule implements SerializableInterface
{
    public function __construct(
        public readonly string $rule,
        public readonly string|int|float|array|null $value = null,
        public readonly ?string $message = null,
        public readonly ?string $validator = null,
        public readonly ?array $config = null,
        public readonly ?string $operator = null,
        public readonly ?array $rules = null,
    ) {}

    public function toArray(): array
    {
        $data = ['rule' => $this->rule];

        if ($this->value !== null) {
            $data['value'] = $this->value;
        }
        if ($this->message !== null) {
            $data['message'] = $this->message;
        }
        if ($this->validator !== null) {
            $data['validator'] = $this->validator;
        }
        if ($this->config !== null) {
            $data['config'] = $this->config;
        }
        if ($this->operator !== null) {
            $data['operator'] = $this->operator;
        }
        if ($this->rules !== null) {
            $data['rules'] = array_map(
                fn(ValidationRule $r) => $r->toArray(),
                $this->rules
            );
        }

        return $data;
    }
}

File: src/ReactUbiquitous.NuGet/Models/Elements/ValidationRule.cs
using System.Text.Json.Serialization;

namespace ReactUbiquitous.NuGet.Models.Elements;

/// <summary>Represents a validation rule applied to a form element.</summary>
public class ValidationRule
{
    /// <summary>Gets or sets the rule type identifier.</summary>
    [JsonPropertyName("rule")]
    public string Rule { get; set; } = string.Empty;

    /// <summary>Gets or sets the value associated with the rule.</summary>
    [JsonPropertyName("value")]
    public object? Value { get; set; }

    /// <summary>Gets or sets the error message to display when the rule fails.</summary>
    [JsonPropertyName("message")]
    public string? Message { get; set; }

    /// <summary>Gets or sets the name of a custom validator function.</summary>
    [JsonPropertyName("validator")]
    public string? Validator { get; set; }

    /// <summary>Gets or sets additional configuration for the rule.</summary>
    [JsonPropertyName("config")]
    public object? Config { get; set; }

    /// <summary>Gets or sets the logical operator used to combine nested rules.</summary>
    [JsonPropertyName("operator")]
    public string? Operator { get; set; }

    /// <summary>Gets or sets a list of nested validation rules.</summary>
    [JsonPropertyName("rules")]
    public List<ValidationRule>? Rules { get; set; }
}
